### PR TITLE
Code generalization re 32/64 bit flexibility.

### PR DESCRIPTION
**Pull request by chipbarnaby**
## Description

Various changes to generalize code to build OK in both 32 and 64 bit.

- Changed many LI (and ULI) to INT (UINT).  INT is always 32 bit.  LI is 32 or 64.  For general working integers, 32 is sufficient.  Some changed to size_t as appropriate (size_t is 32 or 64 bits).  Some internal (non-cnrecs.def) code changed to built-in type int.  File offsets changed to built-in type long to match library IO function arg types.
- Eliminated many INT etc. casts in printf-ish args.  SIs are now implicitly promoted to 32 bits, cast is unnecessary.
- Changed some char * to const char *.  More needed.
- Eliminated comments on some #includes that attempted to document why the include was needed.  These comments are a maintenance hassle and thus generally unreliable.  For example, 
>           #include "cnglob.h"  // LI

- Generalized expression code to be (more) aware of INTs.  Added TYINT re internal data type tracking.
- Generalized cvpak.cpp conversion code re data type changes.  Reworked limit checking code.
- Eliminated some C-style top-of-function declarations w/o initialization in favor of declaration-with-initialization within appropriate scope.
- Changed some integer types (e.g. SI) to bool in boolean contexts.  Need to more fully support bool.
- Clarified use of "long" in comments -- stated actual bit size or built-in type as appropriate.
- Temporary regression testing framework for 64-bit build
-

### CODE REVIEW

File: src/cnguts.cpp [hunk 1]
```diff
@@ -1408,23 +1408,23 @@ ZNR* ZNRES::zr_GetZone() const		// zone of ZNRES
 {	return ZrB.GetAtSafe( ss);
 }		// ZNRES::zr_GetZone
 //-----------------------------------------------------------------------------
-LI ZNRES::zr_GetRunTotalLI( int fn) const
-// return total of ZNRES LI values = cummulative count for run so far
+int ZNRES::zr_GetRunTotalInt( int fn) const
```
Comment (nealkruis) [inline src/cnguts.cpp:1411]: This is hungarian-esque. Can we just call it `zr_GetRunTotal`?

File: src/cueval.cpp [hunk 2]
```diff
@@ -157,20 +157,22 @@ RC FC cuEvalR( 		// evaluate pseudocode & return ptr to value
 	rc = cuEval( ip, pmsg, pBadH);	// evaluate to *evStk
 	if (ppv != NULL)				// return pointer to value in eval stack
 		*ppv = evSp;
+#if 0	// TODO64
```
Comment (nealkruis) [inline src/cueval.cpp:160]: Does this need to be resolved before merge?

File: src/cueval.cpp [hunk 3]
```diff
@@ -882,25 +908,35 @@ w	 case PSRATLOD1S: POINT; *--SPI = (SI)*(CH*)v; break; 	// 1 byte, extend sign
...
+			if ((rc = cuRmGet(vND,&ms,pBadH)) != RCOK)		// char *.  1st fetch/check/fix 4 bytes.
 				goto breakbreak;				//   if unset data or uneval'd expr, ms set.
+#if 0
 			if (p)
 				cupIncRef( DMPP( p));				//   ++ ref count of ptr if not into code (below)
 			else
 				p = strsave("");				//   supply dm "" for NULL pointer
-			*--SPP = p;					//   stack pointer
-			break;
 #endif
```
Comment (nealkruis) [inline src/cueval.cpp:920]: Remove?

File: src/cueval.cpp [hunk 4]
```diff
@@ -882,25 +908,35 @@ w	 case PSRATLOD1S: POINT; *--SPI = (SI)*(CH*)v; break; 	// 1 byte, extend sign
...
+#if 1
+			if ((rc = cuRmGet( vND, &ms, pBadH)) != RCOK)	// 4 bytes: float/INT/UINT.  1st fetch/check/fix 4 bytes.
+				goto breakbreak;		//   if unset data or uneval'd expr, ms set.
+			*--SPL = vND;				//   push value fetched
+#else
+			if ((rc = cuRmGet(&p,&ms,pBadH)) != RCOK)	// 4 bytes: float/INT/UINT.  1st fetch/check/fix 4 bytes.
 				goto breakbreak;			//   if unset data or uneval'd expr, ms set.
-			*--SPP = p;				//   push value fetched
+			*--SPP = p;		//   push value fetched
+#endif
```
Comment (nealkruis) [inline src/cueval.cpp:935]: Another unresolved `#if 0/1`.

File: src/cueval.cpp [hunk 5]
```diff
@@ -1147,6 +1183,8 @@ LOCAL RC FC cuRmGet(
 		}
 	}
 #if 1
+	vRet = v;
+#elif 0
```
Comment (nealkruis) [inline src/cueval.cpp:1187]: Another unresolved `#if 0/1`.

File: src/cuparse.cpp [hunk 6]
```diff
@@ -1009,13 +1009,14 @@ RC FC exOrk(	// compile expression from current input file, return constant valu
 
 // determine if constant
 	void* pv;
-	Eer( konstize( &isK, &pv, 0 ) )	// evals if evaluable and un-eval'd. Rets flag and ptr (ptr to ptr for TYSTR). below.
+	SI isKon{ 0 };
```
Comment (nealkruis) [inline src/cuparse.cpp:1012]: Inching towards readability 😄.

File: src/cuparse.cpp [hunk 7]
```diff
@@ -3482,6 +3520,9 @@ LOCAL RC FC cnvPrevSf( 	// append (conversion) operation to ith previous express
 				if (sz==4)
 				{
 					op1 = PSKON4;     // differences for 4-byte value choicn
+#if CSE_ARCH != 32
+					printf("\nFix cnvPrevSf shift");	// TODO64 shift looks dubious on 64 bit
+#endif
```
Comment (nealkruis) [inline src/cuparse.cpp:3525]: Does this need to be resolved?

File: src/cuprobe.cpp [hunk 8]
```diff
@@ -603,7 +609,7 @@ LOCAL RC FC lopNty4dt( 	// for DT- data type, get TY- type and PSOP to load it f
 		break;
 
 	case DTLDATETIME:			// show dateTime as number (?)
-	case DTLI:
+	case DTLI:		// TODO64 probably don't want DTLI here
```
Comment (nealkruis) [inline src/cuprobe.cpp:612]: Resolve?

File: src/cvpak.h [hunk 9]
```diff
@@ -82,7 +82,7 @@ RC FC cvS2Choi( const char *s, USI dt, void *pv, USI *pSz, const char** pms);
 
 RC FC cvatof( const char *_str, double *vp, bool percent=false );
 #if 0
-x RC FC cvatol(char* s, LI* pn, SI hexoct); 		// convert string to long integer
+x RC FC cvatol(char* s, INT* pn, SI hexoct); 		// convert string to long integer
 x RC FC cvatof2( char *s1, char *s2, double *vp, SI percent );
 #endif
```
Comment (nealkruis) [inline src/cvpak.h:87]: Just remove?

File: src/cueval.cpp [hunk 10]
```diff
@@ -258,6 +256,7 @@ LOCAL RC FC cuEvalI(
 			USI h;
 			void* tIp;
 			void* tSp;
+			NANDAT vND;
 #if 0
 #if 0 && defined( ND3264)
 made things worse  TODO (MP)
```
Comment (nealkruis) [inline src/cueval.cpp:262]: Clean up?

File: src/exman.cpp [hunk 11]
```diff
@@ -764,7 +764,7 @@ RC FC exPile(		// compile an expression from current input
 		EXTDUMPIF( "exPile new end");
 	}
 
-#if 0 && defined( _DEBUG)
+#if 1 && defined( _DEBUG)
```
Comment (nealkruis) [inline src/exman.cpp:767]: remove `1 && `?

File: src/cnglob.h [hunk 12]
```diff
@@ -338,26 +335,11 @@ template< typename T> T& IvlData( T* ivlData, int ivl)
...
-#define ISNANDLE(nandat) (((*reinterpret_cast<const NANDAT *>(&nandat)) & 0xffff0000L)==0xff800000L)	// test for ref to non-constant expr (or unset)
-#define ISNANDLEP(pNandat) ((*(reinterpret_cast<const UI*>(pNandat)) & 0xffff0000L)==0xff800000L)	// test for ptr to ref to non-constant expr (or unset)
-#define EXN(nandle)  ((*reinterpret_cast<const NANDAT *>(&nandle)) & 0xffff)				// extract expression # from nandle
-#define UNSET (NANDLE(0))					// "unset" value for float/ptr/LI.  cast as desired.
-#define ASING (NANDLE(0xffff))				// may be stored in values to be determined by autosizing 6-95
-#define NANDLE(h) (static_cast<NANDAT>(0xff800000 + h))		// "expr n" ref for float/ptr/LI (or SI in 4 bytes). h = 1..16383.
-#elif 1
-// transition aid: 32 bit using AsNANDAT
+// type to hold a NANDLE or a datum
 #define ND3264		// #define to enable general 32/64 bit code  TODO (MP)
```
Comment (nealkruis) [inline src/cnglob.h:339]: Is this still needed?

File: src/cuparse.cpp [hunk 13]
```diff
@@ -1009,13 +1009,14 @@ RC FC exOrk(	// compile expression from current input file, return constant valu
...
-	Eer( konstize( &isK, &pv, 0 ) )	// evals if evaluable and un-eval'd. Rets flag and ptr (ptr to ptr for TYSTR). below.
+	SI isKon{ 0 };
+	Eer( konstize( &isKon, &pv, 0 ) )	// evals if evaluable and un-eval'd. Rets flag and ptr (ptr to ptr for TYSTR). below.
 	USI codeSize;
 	Eer( finPile( &codeSize) )		// now terminate compilation / get size
 
 // for constant, return value and no code
-	ULI v = 0;
-	if (isK)   						// if konstize found (or made) a constant value
+	ULI v = 0;		// TODO64 -- ULI?
```
Comment (nealkruis) [inline src/cuparse.cpp:1018]: Does this need to be addressed still?

File: src/cuparse.cpp [hunk 14]
```diff
@@ -1058,9 +1059,9 @@ RC FC exOrk(	// compile expression from current input file, return constant valu
 	if (rc != RCOK)
 	{
 er:    // Eer macro comes here on non-RCOK fcn return
-		isK = 1;  			// say is constant (ret *pisK)
+		isKon = 1;  			// say is constant (ret *pisKon)
 		if (gotTy==TYSTR)		// gotTy init to wanTy above
-			v = (ULI)"";  		// for string return null string 
+			v = (ULI)"";  		// for string return null string TODO64 -- ULI?
```
Comment (nealkruis) [inline src/cuparse.cpp:1064]: TODO when?

File: src/cuparse.cpp [hunk 15]
```diff
@@ -1071,13 +1072,13 @@ RC FC exOrk(	// compile expression from current input file, return constant valu
...
-		*pisK = isK;      		// non-0 if already-evaluated constant
+	if (pisKon)
+		*pisKon = isKon;      		// non-0 if already-evaluated constant
 	if (pip)
 		*pip = ip;			// NULL or pointer to pseudo-code
-	if (isK)				// *pvPar left unchanged if non-constant
+	if (isKon)				// *pvPar left unchanged if non-constant
 		if (pvPar)
-			*(ULI*)pvPar = v;		// value
+			*(NANDAT*)pvPar = v;		// TODO64: was ULI*, NANDAT* seems right
```
Comment (nealkruis) [inline src/cuparse.cpp:1081]: TODO when?

File: src/cuparse.cpp [hunk 16]
```diff
@@ -3237,7 +3268,7 @@ LOCAL SI FC isKE(		// test if *parSp is a constant expression
 	    2: constant expression, but not last subexpression emitted
 	    3: constant expression that can be evaluated now and replaced in place with a constant */
 {
-	USI sz, szkon;   PSOP kop;   static char *p;
+	static char *p;		// TODO64 -- yikes, return value in static?  See return 1 case
```
Comment (nealkruis) [inline src/cuparse.cpp:3271]: TODO when?

File: src/CNRECS.DEF [hunk 17]
```diff
@@ -272,7 +272,7 @@ RECORD TOPRAT "top" *RAT	/* top level RAT: contains control info and all once-on
 							//   C_VENTAVAILCH_WHOLEHOUSE: airnet windows / fans
 							//   C_VENTAVAILCH_ZONAL: airnet windows / fans (controlled by zone)
 							//   C_VENTAVAILCH_RSYSOAV: RSYS outside air vent (central mechanical)
-	*declare "ULI tp_GetVentAvail() const { return CHN( tp_ventAvail); }"
+	*declare "INT tp_GetVentAvail() const { return CHN( tp_ventAvail); }"
```
[the remark this answers is not in the document]
Comment (chipbarnaby) [reply src/CNRECS.DEF:275]: Agree that policy is vague.  I have done some cleanup, generally moving toward native C++ in non-record-members.

[the remark this answers is not in the document]
Comment (chipbarnaby) [reply src/CNRECS.DEF:275]: Agree policy is vague.  Have done some cleanup moving toward native C++ for non-record members

File: src/cnglob.h (hunk 12 above)
Comment (chipbarnaby) [reply src/cnglob.h:339]: No, dropped.

File: src/cueval.cpp (hunk 2 above)
Comment (chipbarnaby) [reply src/cueval.cpp:160]: Done

File: src/cueval.cpp (hunk 3 above)
Comment (chipbarnaby) [reply src/cueval.cpp:920]: Cleaned up

File: src/cueval.cpp (hunk 4 above)
Comment (chipbarnaby) [reply src/cueval.cpp:935]: Cleaned up

File: src/cueval.cpp (hunk 5 above)
Comment (chipbarnaby) [reply src/cueval.cpp:1187]: Fixed

**Review by nealkruis — APPROVED** [review]

**Pull request merged**